File: Backtracking/sudoku/sudoku.js
// Write a program to solve a Sudoku puzzle by filling the empty cells.

// A sudoku solution must satisfy all of the following rules:

// Each of the digits 1-9 must occur exactly once in each row.
// Each of the digits 1-9 must occur exactly once in each column.
// Each of the digits 1-9 must occur exactly once in each of the 9 3x3 sub-boxes of the grid.
// The '.' character indicates empty cells.
// https://leetcode.com/problems/sudoku-solver/

// contents of lesson
// Step 1. verify the constraints
// What happens if the board cannot be solved?
// > Leave the sudoku puzzle as is.

// Step2. Write out some test cases

/**
 * Backtracking Template
 * inside recursive function,
 * 1. add
 * 2. decide
 * 3. if valid call recursive func
 * 4. remove
 */



function getBoxId(row, col) {
  return Math.floor(row / 3) + Math.floor(col / 3) * 3
}

function sudokuSolver(sudoku) {
  const N = sudoku[0].length;
  const EMPTY_STR = '.';
  const rows = new Array(N).fill(EMPTY_STR).map(() => {});
  const cols = new Array(N).fill(EMPTY_STR).map(() => {});
  const boxes = new Array(N).fill(EMPTY_STR).map(() => {});

  recursive(0, 0, 0);

  function isValid(r, c, value) {
    return (r >= 0 && r < N && c >= 0 && c < N)
      && (!rows[r][value] && !cols[c][value] && !boxes[getBoxId(r, c)][value])
  }

  function recursive(r, c, value) {
    if (value >= N || r >= N || c >= N) {
      return;
    }

    if (sudoku[r][c] === EMPTY_STR) {
      sudoku[r][c] = value;
    }
    if (isValid(r, c, value)) {
      rows[r][value] = true;
      cols[c][value] = true;
      boxes[getBoxId(r, c)][value] = true;

      value = value === (N - 1)? 0 : (value + 1);
      if (c === N - 1) {
        c = 0, r += 1;
      } else {
        c += 1;
      }
    } else {
      sudoku[r][c] = EMPTY_STR;
      value = value === 0? N - 1 : (value - 1);
    }
    recursive(r, c, value);
  }

  return sudoku;

}

File: Backtracking/sudoku/python/sudoku.py
def solutions(board):
    def get_box_id(row, col):
        return (row // 3) * 3 + (col // 3)

    def is_safe(row, col, val):
        return not (val in rows[row] or val in cols[col] or val in boxes[get_box_id(row, col)])

    def solve(row, col):
        if row == len(board) or col == len(board):
            return True
        if board[row][col] == '.':
            for i in range(1, 10):
                if is_safe(row, col, i):
                    board[row][col] = str(i)
                    rows[row].add(i)
                    cols[col].add(i)
                    boxes[get_box_id(row, col)].add(i)
                    if col == len(board) - 1:
                        if solve(row + 1, 0):
                            return True
                    else:
                        if solve(row, col + 1):
                            return True
                    board[row][col] = '.'
                    rows[row].remove(i)
                    cols[col].remove(i)
                    boxes[get_box_id(row, col)].remove(i)
        else:
            if col == len(board) - 1:
                if solve(row + 1, 0):
                    return True
            else:
                if solve(row, col + 1):
                    return True
        return False

    rows = [set() for _ in range(len(board))]
    cols = [set() for _ in range(len(board))]
    boxes = [set() for _ in range(len(board))]

    for i in range(len(board)):
        for j in range(len(board)):
            if board[i][j] != '.':
                rows[i].add(int(board[i][j]))
                cols[j].add(int(board[i][j]))
                boxes[get_box_id(i, j)].add(int(board[i][j]))

    solve(0,0)
    return board

board = [
  ['5', '3', '.', '.', '7', '.', '.', '.', '.'],
  ['6', '.', '.', '1', '9', '5', '.', '.', '.'],
  ['.', '9', '8', '.', '.', '.', '.', '6', '.'],
  ['8', '.', '.', '.', '6', '.', '.', '.', '3'],
  ['4', '.', '.', '8', '.', '3', '.', '.', '1'],
  ['7', '.', '.', '.', '2', '.', '.', '.', '6'],
  ['.', '6', '.', '.', '.', '.', '2', '8', '.'],
  ['.', '.', '.', '4', '1', '9', '.', '.', '5'],
  ['.', '.', '.', '.', '8', '.', '.', '7', '9']
]
print(solutions(board))

        